src/telem/lib/output/plotter/ChartMultiLineOutputAdapter.lua

```diff
@@ -0,0 +1,202 @@
+local MetricCollection  = require 'telem.lib.MetricCollection'
+
+local ChartMultiLineOutputAdapter = o.class(OutputAdapter)
+ChartMultiLineOutputAdapter.type = 'ChartMultiLineOutputAdapter'
+
+ChartMultiLineOutputAdapter.MAX_ENTRIES = 50
+ChartMultiLineOutputAdapter.X_TICK = 10
+
+function ChartMultiLineOutputAdapter:constructor (win, filter, bg, fg, maxEntries)
+    self:super('constructor')
+
+    self:cacheable()
+
+    self.win = assert(win, 'Window is required')
+    self.filter = assert(filter, 'Filter is required')
+
+    for i, metric in ipairs(filter) do
+        assert(metric.name, 'Metric #' .. i .. ' name is required')
+        assert(metric.color, 'Metric #' .. i .. ' color is required')
+    end
+
+    self.plotter = nil
+    self.plotData = {}
+    for i,v in ipairs(filter) do
+        self.plotData[i] = {}
+    end
+    self.gridOffsetX = 0
+
+    self.filter = filter
+
+    self.bg = bg or win.getBackgroundColor() or colors.black
+    self.fg = fg or win.getTextColor() or colors.white
+    self.MAX_ENTRIES = maxEntries or self.MAX_ENTRIES
+
+    self:register()
+end
+
+function ChartMultiLineOutputAdapter:register ()
+    if not vendor then
+        self:dlog('ChartMultiLineOutputAdapter:boot :: Loading vendor modules...')
+
+        vendor = require 'telem.vendor'
+
+        self:dlog('ChartMultiLineOutputAdapter:boot :: Vendor modules ready.')
+    end
+
+    if not plotterFactory then
+        self:dlog('ChartMultiLineOutputAdapter:boot :: Loading plotter...')
+
+        plotterFactory = vendor.plotter
+
+        self:dlog('ChartMultiLineOutputAdapter:boot :: plotter ready.')
+    end
+
+    self:updateLayout()
+
+    for i=1, #self.plotData do
+        for j=1, self.MAX_ENTRIES do
+            t.constrainAppend(self.plotData[i], self.plotter.NAN, self.MAX_ENTRIES)
+        end
+    end
+end
+
+function ChartMultiLineOutputAdapter:updateLayout (bypassRender)
+    self.plotter = plotterFactory(self.win)
+
+    if not bypassRender then
+        self:render()
+    end
+end
+
+function ChartMultiLineOutputAdapter:write (collection)
+    assert(o.instanceof(collection, MetricCollection), 'Collection must be a MetricCollection')
+
+    for i, metric in ipairs(self.filter) do
+        local resultMetric = collection:find(metric.name)
+        assert(resultMetric, 'could not find metric')
+
+        -- TODO data width setting
+        self.gridOffsetX = self.gridOffsetX - t.constrainAppend(self.plotData[i], resultMetric and resultMetric.value or self.plotter.NAN, self.MAX_ENTRIES)
+
+        -- TODO X_TICK setting
+        if self.gridOffsetX % self.X_TICK == 0 then
+            self.gridOffsetX = 0
+        end
+
+        -- lazy layout update
+        local winw, winh = self.win.getSize()
+        if winw ~= self.plotter.box.term_width or winh ~= self.plotter.box.term_height then
+            self:updateLayout(true)
+        end
+    end
+
+    self:render()
+    
+    return self
+end

```

Comment:
This implementation is resulting in the background grid shifting once for every plot series, rather than once per cycle, causing a parallax effect. The code below should fix the issue by moving the grid calculation math outside of the series loop.

```lua
function ChartMultiLineOutputAdapter:write (collection)
    assert(o.instanceof(collection, MetricCollection), 'Collection must be a MetricCollection')

    local maxDataRemoved = 0

    for i, metric in ipairs(self.filter) do
        local resultMetric = collection:find(metric.name)
        assert(resultMetric, 'could not find metric')

        -- TODO data width setting
        maxDataRemoved = math.max(maxDataRemoved, t.constrainAppend(self.plotData[i], resultMetric and resultMetric.value or self.plotter.NAN, self.MAX_ENTRIES))
    end

    self.gridOffsetX = self.gridOffsetX - maxDataRemoved

    -- TODO X_TICK setting
    if self.gridOffsetX % self.X_TICK == 0 then
        self.gridOffsetX = 0
    end

    -- lazy layout update
    local winw, winh = self.win.getSize()
    if winw ~= self.plotter.box.term_width or winh ~= self.plotter.box.term_height then
        self:updateLayout(true)
    end

    self:render()
    
    return self
end
```

Reply:
Yeah, this is much better, thank you!